### PR TITLE
(feat) Implement configurable enabledErrorTypes

### PR DESCRIPTION
**Pull request by robinmacharg**
## Goal

To allow users to configure the types of event that are reported.

## Design

This is implemented as a bitfield on `BugsnagConfiguration`.  `BugsnagSentry` translates this to equivalent KSCrash values. The OOM value is examined in `BugsnagNotifier`. The bitfield mapping has been broken out to make it testable.

## Changeset

As above, `BugsnagConfiguration`, `BugsnagSentry` and `BugsnagNotifier`.

## Tests

This is difficult to test at unit-level.  What can be has been.  E2E tests better test the behaviour and this is where the effort has been concentrated.

### Outstanding Questions

* E2E Testing is extensive but I have lingering doubts over whether it's __comprehensive__.  There's ambiguous behaviour around e.g. Mach errors being converted to Signals.

- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [X] Final review

- [ ] The correct target branch has been selected (`master` for fixes, `next` for
  features)
- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [X] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [X] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

**Comment by robinmacharg** [comment]
Note to self: I need to check e2e on Mac vs iOS.

File: Source/BugsnagConfiguration.m [hunk 1]
```diff
@@ -336,4 +342,6 @@ - (void)setMaxBreadcrumbs:(NSUInteger)capacity {
     self.breadcrumbs.capacity = capacity;
 }
 
+@synthesize enabledErrorTypes = _enabledErrorTypes;
```
Comment (kattrali) [inline Source/BugsnagConfiguration.m]: This should not be necessary since there's no overridden setter and getter.

File: Source/BugsnagNotifier.m [hunk 2]
```diff
@@ -399,14 +399,19 @@ - (void)start {
 
     _started = YES;
     // autoDetectErrors disables all unhandled event reporting
-    BOOL configuredToReportOOMs = self.configuration.reportOOMs && self.configuration.autoDetectErrors;
+    BOOL configuredToReportOOMs = self.configuration.reportOOMs
```
Comment (kattrali) [inline Source/BugsnagNotifier.m]: The new error types should replace `reportOOMs`, right?

File: UPGRADING.md [hunk 3]
```diff
@@ -45,6 +45,8 @@ Swift:
 - config.add(beforeSession:)
 + config.onSessionBlocks
 + config.add(onSession:)
+
++ config.enabledErrorTypes
```
Comment (kattrali) [inline UPGRADING.md]:
```suggestion
- config.reportOOMs
+ config.enabledErrorTypes
```

File: Source/BugsnagConfiguration.m [hunk 4]
```diff
@@ -106,6 +106,12 @@ -(instancetype)initWithApiKey:(NSString *)apiKey
     _breadcrumbs = [BugsnagBreadcrumbs new];
     _automaticallyCollectBreadcrumbs = YES;
     _autoTrackSessions = YES;
+    // Default to recording all error types
+    _enabledErrorTypes = BSGErrorTypesOOMs
```
Comment (kattrali) [inline Source/BugsnagConfiguration.m]: Enabling OOM detection by default should be guarded to only happen in release builds, to avoid triggering the heuristic when killing/restarting an app in Xcode or similar.

File: Source/BugsnagNotifier.m (hunk 2 above)
Comment (robinmacharg) [reply Source/BugsnagNotifier.m]: I wasn't sure.  Do we want to deprecate this in the first instance rather than excise it completely?

Comment (kattrali) [reply Source/BugsnagNotifier.m]: That would be fine too, though the setter/getter would need to proxy to enabled error types rather than leaving both in place. In the current configuration, there isn't a way to determine precedence since the property can't be unset and is off by default in some cases.

Comment (robinmacharg) [reply Source/BugsnagNotifier.m]: Excised as an exercise in simplicity.
